herddb-core/src/main/java/herddb/core/TableSpaceManager.java

```diff
@@ -1530,79 +1535,85 @@ TableSpaceCheckpoint checkpoint(boolean full, boolean pin, boolean alreadLocked)
+                            if (afterTableCheckPointAction != null) {
+                                afterTableCheckPointAction.run();
+                            }
                         }
                     }
                 }
-            }
 
-            // we are sure that all data as been flushed. upon recovery we will replay the log starting from this position
-            actions.addAll(dataStorageManager.writeCheckpointSequenceNumber(tableSpaceUUID, logSequenceNumber));
+                // we are sure that all data as been flushed. upon recovery we will replay the log starting from this position
+                actions.addAll(dataStorageManager.writeCheckpointSequenceNumber(tableSpaceUUID, logSequenceNumber));
 
-            /* Indexes checkpoint is handled by TableManagers */
-            if (leader) {
-                log.dropOldLedgers(logSequenceNumber);
-            }
+                /* Indexes checkpoint is handled by TableManagers */
+                if (leader) {
+                    log.dropOldLedgers(logSequenceNumber);
+                }
 
-            _logSequenceNumber = log.getLastSequenceNumber();
-        } finally {
-            if (!alreadLocked) {
-                releaseWriteLock(lockStamp, "checkpoint");
+                _logSequenceNumber = log.getLastSequenceNumber();
+            } finally {
+                if (!alreadLocked) {
+                    releaseWriteLock(lockStamp, "checkpoint");
+                }
             }
-        }
 
-        for (PostCheckpointAction action : actions) {
-            try {
-                action.run();
-            } catch (Exception error) {
-                LOGGER.log(Level.SEVERE, "postcheckpoint error:" + error, error);
+            for (PostCheckpointAction action : actions) {
+                try {
+                    action.run();
+                } catch (Exception error) {
+                    LOGGER.log(Level.SEVERE, "postcheckpoint error:" + error, error);
+                }
             }
-        }
-        long _stop = System.currentTimeMillis();
 
-        LOGGER.log(Level.INFO, "{0} checkpoint finish {1} started ad {2}, finished at {3}, total time {4} ms",
-                new Object[]{nodeId, tableSpaceName, logSequenceNumber, _logSequenceNumber, Long.toString(_stop - _start)});
-
-        return new TableSpaceCheckpoint(logSequenceNumber, checkpointsTableNameSequenceNumber);
+        } finally {
+            long _stop = System.currentTimeMillis();
+            LOGGER.log(Level.INFO, "{0} checkpoint finish {1} started ad {2}, finished at {3}, total time {4} ms",
+                    new Object[]{nodeId, tableSpaceName, logSequenceNumber, _logSequenceNumber, Long.toString(_stop - _start)});
+            statsLogger.getOpStatsLogger(this.tableSpaceName).registerSuccessfulEvent(_stop, TimeUnit.MILLISECONDS);
+            return new TableSpaceCheckpoint(logSequenceNumber, checkpointsTableNameSequenceNumber);
```

Comment:
It is not good to have a 'return' inside a finally block, please leave it inside the main try block